File: rust/src/encoding/general.rs
use std::{collections::HashMap, hash::Hash};

use crate::encoding::nb::{decode_vuint, encode_vuint};

// in_field = true, omit length field
#[inline]
pub fn encode_bool(data: &mut Vec<u8>, value: bool) {
	data.push(value as u8);
}
#[inline]
pub fn decode_bool(data: &[u8], ind: &mut usize) -> Option<bool> {
	let value = *data.get(*ind)? != 0;
	*ind += 1;
	Some(value)
}

#[inline]
pub fn encode_u8_arr(data: &mut Vec<u8>, value: &[u8]) {
	encode_vuint(data, value.len() as u64);
	data.extend_from_slice(value);
}
#[inline]
pub fn decode_u8_arr(data: &[u8], ind: &mut usize) -> Option<Vec<u8>> {
	let len = decode_vuint(data, ind)? as usize;
	let value = data.get(*ind..*ind + len)?.to_vec();
	*ind += len;
	Some(value)
}

#[inline]
pub fn encode_str(data: &mut Vec<u8>, value: &str) {
	encode_u8_arr(data, value.as_bytes());
}
#[inline]
pub fn decode_str(data: &[u8], ind: &mut usize) -> Option<String> {
	Some(String::from_utf8(decode_u8_arr(data, ind)?).ok()?)
}

#[inline]
pub fn encode_arr<T>(
	data: &mut Vec<u8>, value: &[T], in_field: bool, item_fn: impl Fn(&mut Vec<u8>, &T) -> (),
) {
	if !in_field {
		encode_vuint(data, value.len() as u64)
	}
	for v in value {
		item_fn(data, v);
	}
}
#[inline]
pub fn decode_arr<T>(
	data: &[u8], ind: &mut usize, in_field: bool, item_fn: impl Fn(&[u8], &mut usize) -> Option<T>,
) -> Option<Vec<T>> {
	let len = decode_vuint(data, ind)? as usize;
	if in_field {
		let start_ind = *ind;
		let mut vec = Vec::new();
		while *ind < start_ind + len {
			vec.push(item_fn(data, ind)?);
		}
		Some(vec)
	} else {
		let mut vec = Vec::with_capacity(len);
		for _ in 0..len {
			vec.push(item_fn(data, ind)?);
		}
		Some(vec)
	}
}

#[inline]
pub fn encode_map<K, V>(
	data: &mut Vec<u8>, value: &HashMap<K, V>, in_field: bool,
	key_fn: impl Fn(&mut Vec<u8>, &K) -> (), val_fn: impl Fn(&mut Vec<u8>, &V) -> (),
) {
	if !in_field {
		encode_vuint(data, value.len() as u64)
	}
	for (k, v) in value {
		key_fn(data, k);
		val_fn(data, v);
	}
}
#[inline]
pub fn decode_map<K: Eq + Hash, V>(
	data: &[u8], ind: &mut usize, in_field: bool, key_fn: impl Fn(&[u8], &mut usize) -> Option<K>,
	val_fn: impl Fn(&[u8], &mut usize) -> Option<V>,
) -> Option<HashMap<K, V>> {
	let len = decode_vuint(data, ind)? as usize;
	if in_field {
		let start_ind = *ind;
		let mut map = HashMap::new();
		while *ind < start_ind + len {
			let k = key_fn(data, ind)?;
			let v = val_fn(data, ind)?;
			map.insert(k, v);
		}
		Some(map)
	} else {
		let mut map = HashMap::with_capacity(len);
		for _ in 0..len {
			let k = key_fn(data, ind)?;
			let v = val_fn(data, ind)?;
			map.insert(k, v);
		}
		Some(map)
	}
}

File: js/src/general.ts
import { decode_u8, decode_u8_arr, encode_u8, encode_u8_arr, reserve, type Buffer, type Cursor } from "./buf.ts";
import { decode_vuint, encode_vuint } from "./number.ts";

export function encode_bool(buf: Buffer, value: boolean) {
	encode_u8(buf, value === true ? 1 : 0)
}
export function decode_bool(buf: Buffer, cur: Cursor) {
	return decode_u8(buf, cur) !== 0
}

let TextEnc = new TextEncoder(); 
export function encode_str(buf: Buffer, value: string) {
	let res = TextEnc.encode(value);
	encode_vuint(buf, res.length);
	encode_u8_arr(buf, res);
}

let TextDec = new TextDecoder('utf-8');
export function decode_str (buf: Buffer, cur: Cursor) {
	let size = decode_vuint(buf, cur) as number;
	return TextDec.decode(decode_u8_arr(buf, size, cur));
}

type Encoder <T> = (buf: Buffer, value: T) => void;
type Decoder <T> = (buf: Buffer, cur: Cursor) => T;

export function encode_arr<T> (buf: Buffer, value: T[], item_fn: Encoder<T>, in_field = false) {
	if (!in_field) 
		encode_vuint(buf, value.length);
	for (let val of value) 
		item_fn(buf, val);
}
export function decode_arr<T> (buf: Buffer, cur: Cursor, item_fn: Decoder<T>, in_field = false) {
	let len = decode_vuint(buf, cur) as number;
	if (in_field) {
		let start_ind = cur.pos;
		let arr: T[] = [];
		while (cur.pos < start_ind + len)
			arr.push(item_fn(buf, cur));
		return arr
	} else {
		let arr: T[] = [];
		for (let i = 0; i < len; i++) 
			arr.push(item_fn(buf, cur));
		return arr
	}
}

export function encode_map<K, V> (
	buf: Buffer, value: Map<K, V>, key_fn: Encoder<K>, val_fn: Encoder<V>, 
	in_field = false
) {
	if (!in_field) 
		encode_vuint(buf, value.size);
	for (let [k, v] of value) {
		key_fn(buf, k);
		val_fn(buf, v);
	}
}
export function decode_map<K, V> (
	buf: Buffer, cur: Cursor, key_fn: Decoder<K>, val_fn: Decoder<V>, 
	in_field = false
) {
	let len = decode_vuint(buf, cur) as number;
	if (in_field) {
		let start_ind = cur.pos;
		let map = new Map<K, V>();
		while (cur.pos < start_ind + len) 
			map.set(key_fn(buf, cur), val_fn(buf, cur));
		return map
	} else {
		let map = new Map<K, V>();
		for (let i = 0; i < len; i++) {
			map.set(key_fn(buf, cur), val_fn(buf, cur));}
		return map
	}
}

export function skip_field (buf: Buffer, cur: Cursor, header: number) {
	switch (header & 0b111) {
		case 0b000: cur.pos += 1; break;
		case 0b001: cur.pos += 2; break;
		case 0b010: cur.pos += 4; break;
		case 0b011: cur.pos += 8; break;
		// decode vuint and ignore
		case 0b100: decode_vuint(buf, cur); break;
		// len field is encoded
		case 0b101: cur.pos += decode_vuint(buf, cur) as number;
	}
}